### ISSUE TITLE
Expose nl_tape capabilities to Python: in-memory construction, HVP, and an erf TapeOp

### ISSUE THREAD

**Issue by jkitchin** [jkitchin/pounce#469]
## Context

[discopt](https://github.com/jkitchin/discopt) is moving its NLP subproblem derivatives off JAX and onto pounce's Rust AD tape (`pounce-nl/src/nl_tape.rs`). pounce is *already* the NLP subsolver — `discopt/solvers/nlp_pounce.py` builds cyipopt-shaped callbacks and hands them to pounce — but those callbacks are currently computed by a JAX evaluator. The change is a callback-provider substitution: one tape per model, replacing `jax.grad`/`jacfwd`/`hessian`.

Verified this works in principle: `pounce.read_nl()` on a 580-var/510-constraint instance reads in 8 ms and its `objective`/`gradient` agree with the JAX path **exactly** (relative difference 0.00e+00, gradient max-abs-diff 0.0).

Three gaps block the swap. All three are capabilities that already exist in the Rust core and are simply not reachable from Python.

## 1. No in-memory problem construction (biggest blocker)

`pounce-py` exposes only file-path `read_nl` (`crates/pounce-py/src/nl_problem.rs:421`). Confirmed by grep — `crates/pounce-py/src/` has **zero** references to:

- `parse_nl_text` (`crates/pounce-nl/src/nl_reader.rs:322`, `pub fn`)
- `Tape::build` / `Tape::build_multi` (`crates/pounce-nl/src/nl_tape.rs:180`, `:1912`, `pub fn`)

discopt models built through its Python modeling API never touch disk, so today the only route is writing a temporary `.nl` file. That is not just slow — it is **lossy**. discopt's `.nl` writer refuses several operators that pounce's tape *already supports*:

| operator | in pounce `TapeOp`? | survives discopt's `.nl` export? |
|---|---|---|
| `atan2` | yes (`Atan2`) | no — writer has no 2-arg funcall path |
| `min` / `max` | yes (`Min`, `Max`) | no — raises "requires DNLP model type" |

So round-tripping through `.nl` would *lose* capability that the tape has. Either binding solves it; `Tape::build(expr)` is the more direct fit if the expression IR is exposable, otherwise `parse_nl_text(&str)` at least avoids the filesystem.

**Ask:** a pyo3 binding for in-memory construction — `parse_nl_text(text) -> NlProblem`, and/or a tape-builder entry point.

## 2. No Hessian-vector product exposed

`hessian_directional` (`crates/pounce-nl/src/nl_tape.rs:696`) is `pub fn` with zero references in `pounce-py`. `pounce.NlProblem`'s Python surface is:

```
con_names, constraints, g_l, g_u, gradient, hessian, hessian_structure,
jacobian, jacobian_structure, m, minimize, n, nnz_hess, nnz_jac,
obj_constant, objective, var_names, variant, x0, x_l, x_u
```

— no HVP. discopt uses one (`_jax/nlp_evaluator.py:485-493`, a `jax.grad` + `jax.jvp` composition) for large models where the dense Lagrangian Hessian is impractical.

Everything else discopt needs is already exposed and matches well: the Ipopt-style `hessian(x, lam, obj_factor)` plus `hessian_structure`/`jacobian_structure` map directly onto discopt's `evaluate_hessian_values`/`hessian_structure`, and the sparse structures replace discopt's JAX sparsity probe and CPR-coloring helpers outright.

**Ask:** expose `hessian_directional` on `PyNlProblem`.

## 3. `erf` has no `TapeOp`

Diffing discopt's 30 DAG operators against pounce's `TapeOp` variants: **20 covered, 10 missing** — but 9 of the 10 are decomposable into ops pounce already has, and discopt can do that decomposition on its side (it already does exactly this in its `.nl` writer for `log1p`, `sigmoid`, `softplus`):

| missing | decomposition | needs pounce change? |
|---|---|---|
| `log1p`, `log2`, `sigmoid`, `softplus` | `log`/`exp` compositions | no |
| `entropy`, `centropy` | `x·log(x)`, `x·log(x/y)` | no |
| `prod` | `Mul` chain | no |
| `sign`, `signpower` | `Cmp` + `Select` (both present) | no |
| **`erf`** | **not elementary** | **yes** |

`erf` is the only genuine primitive gap. Its derivative is cheap and closed-form:

```
erf'(x) = 2/sqrt(pi) * exp(-x^2)
erf''(x) = -4x/sqrt(pi) * exp(-x^2)
```

**Ask:** an `Erf` `TapeOp` with value/tangent/adjoint arms. If that's unwanted, a loud unsupported-operator error is a fine answer — discopt would then refuse `erf` models on this path rather than silently mis-solve.

## Notes

- Items 1 and 2 are pure bindings for existing `pub fn`s — no behavior change to any existing API.
- Item 3 is the only new numerics.
- Not urgent: the discopt side is staged behind a default-off flag and gated on a differential correctness panel (`incorrect_count = 0`, no bound above its reference optimum) before anything graduates.

For reference, the full derivative surface discopt needs from the tape: objective, gradient, constraints, Jacobian (values + structure), Lagrangian Hessian (values + structure), and HVP. All but the last are already exposed on `PyNlProblem`.

### Suggested order

1. `hessian_directional` binding — smallest, unblocks large-model Hessian-free mode.
2. `parse_nl_text` binding — unblocks the Python-API model path without a `Tape` IR decision.
3. `Tape::build` binding — the better long-term seam; avoids `.nl` as an intermediate entirely.
4. `Erf` TapeOp — only needed for `erf` models; a loud refusal is an acceptable interim.
